README.md

```diff
@@ -0,0 +1,320 @@
+
+The cat & breed repository classes are pre-populated with properties containing cat & breeds objects, respectively. 
+Each class contains getter and setter methods to retrieve and update data as a complete set. 
+The subsequent methods employ the getter and setter methods, manipulating data further depending on the request.
+
+```javascript
+```
+
+## Cat Repository ##
+
+The cat repository takes an idGenerator argumen which generates a unique id for each cat object in the database upon class instantion. 
+Dependancy injection in the constructor allows us to pass in a mock id generator for testing. 
+
+```javascript
+class CatRepository {
+    constructor(idGenerator) { 
+        this.generateId = idGenerator; 
+        ...
+    }
+    ...
+}
+```
+
+getIndexById() receives a cat id as argument. It searches the database and returns the index of the record with matching id. If none found, returns null. This function is called locally to check whether the database contains a record with a given cat id.  
+
+getAllCats() returns a summary object of each cat. 
+
+```javascript
+{   
+                id: this.generateId(),  // TypeError: this.generateId is not a function
+                name: "Catty",
+                description: "buff or tan, skinny, talkative, often found in tall grasses or deep in the bush"
+}
+```
+
+getCatById() receives a cat id as argument. Returns a detail object of the matching cat. If none found, returns null. 
+
+```javascript
+{   
+                id: this.generateId(),  // TypeError: this.generateId is not a function
+                name: "Catty",
+                ageInYears: 1,
+                favouriteToy: "grass",
+                description: "buff or tan, skinny, talkative, often found in tall grasses or deep in the bush",
+                breedId: 1
+}
+```
+updateCatById() receives a cat object and a cat id. It updates only properties which differ from the existing record. Returns the updated cat object. If no matching record found, returns null. 
+
+deleteCatById() receives a cat id. It deletes the matching cat record in the database. If no matching record found, returns null. 
+
+addCat() receives a cat object to be added to the database. It attaches a new unique id to the object, and then updates the database via setter method. Returns the newly-added cat object.
+
+## Validation Functions ##
+
+Non-middleware simple JS functions exist within a separate module from the Router and Repository files:
+
+isInvalidString() checks the data type of a passed argument. Returns a truthy in the event of a non-string. Else, returns falsey. 
+isInvalidNum() checks the data type of a passed argument. Returns a truthy in the event of a non-number. Else, returns falsey. 
+
+generateErr400() receives a string argument of a descriptive error message. Returns an 400 BAD REQUEST error.
+generateErr404() receives a string of an id which has not been found in the database. Returns an 400 NOT FOUND error.
```

Comment:
It would be worth considering moving these to [inline comments](https://stackoverflow.com/a/127106), as this is very detailed and useful but might get lost to anyone reading the code instead of the README. The code could also end up getting out of date with the docs in the README.

Reply:
I assume you meant just those three functions? :)

Reply:
It was more of a general comment. It doesn't matter too much in this case because the level of documentation is amazing, but might be something to watch out for in future.